### ISSUE TITLE
Using negative loglos with ann doesn't seem to be working correctly

### ISSUE THREAD

**Issue by AllIAskOfYou** [mlpack/mlpack#3372]
Hello,

I tried to use mlpacks's ann api to create softmax regression. When presenting the model with the following example:
``` c++
X = [ [0], [0] ], Y = [ [0], [1] ]
```
the model should learn to predict both classes, given the input of 0, with the same probability, but it doesn't. They are way off.
Here is the code to recreate this:
``` c++
mat X(1, 2);
mat Y(1, 2);
Y(0, 1) = 1;

FFN<> model;
model.Add<Linear>(3);
model.Add<Softmax>();

model.Train(X, Y);
mat pred;
model.Predict(X, pred);
cout << pred << endl;
```
I also tried using a different optimizer with different options but the results were the same.
So do you maybe know why would this be happening?

**Comment by rcurtin** [comment]
Did you mean to use `X = [ [0], [1] ]`?  If you take `X = [ [0], [0] ]`, then the data is inseparable and the network can't learn.  I added the line `X(0, 1) = 1` before training and the example worked like you expected.

(Also as a note, you probably want only 2 neurons in the linear layer, since your data has only 2 classes, but it still works either way.)

**Comment by AllIAskOfYou** [comment]
Hello, thank you for your answear, but I actually ment it the way I wrote it.
I know, that the data is inseparable, and in that case the model should learn to predict into both labels with the same probability (0.5, 0.5) or with 3 labels (0.5, 0.5, 0). I also used 3 output newrons intentionally, since using 2 output classes got the predictions close to the expected output, somewhere around (0.47, 0.53), while using 3 output classes lead to predictions that were more off, like (0.27, 0.71, 0.02) in best case.

**Comment by AllIAskOfYou** [comment]
Hello,
@rcurtin, I just want to ask, if this has been recognised as a problem and if it will be looked into.
Thanks

**Comment by rcurtin** [comment]
Thanks for the clarification @AllIAskOfYou.  I haven't had a chance to look into this quite yet (it's a busy couple days) but I'll see if I can fit it in shortly.  Intuitively I see what you mean that ideally the model should give equal-probability predictions.  But, there are many reasons and small details that could explain why an optimizer would simply give up.  So, I'm not yet sure whether or not this is a problem; I need to dig through what is actually happening in the optimization at a detailed level, and see why it is terminating before converging to what you expect.  It may be that slightly different settings for the optimizer, or a different optimizer altogether, might yield the results you want.

**Comment by AllIAskOfYou** [comment]
Thank you for taking the time to answear. I understand. I'd just like to add, that I tried the same thing in tensorflow, with adam optimizer and it worked as expected. Maybe this information can help you in someway.

**Comment by rcurtin** [comment]
It took a while for me to find the time to reproduce the issue correctly.  Overall I think that this issue is the result of the very specific dataset you have chosen, and also the initialization.  I don't think it has to do with the optimizer.  (In fact in some sense I would say that ensmallen's optimizers seem to do a "better" job, but, again, super specific problem so I don't mean to draw any larger conclusion there.)

So, first, here's my reproduction in Python:

```py
import numpy as np
import tensorflow as tf

X = np.zeros((2, 1))
Y = np.zeros((2, 3), dtype=int)
# One-hot encoding is necessary for the categorical cross-entropy loss.
Y[0, 1] = 1
Y[1, 0] = 1

model = tf.keras.models.Sequential([
    tf.keras.layers.Dense(3),
    tf.keras.layers.Softmax()
])

model.compile(optimizer='adam', loss=tf.keras.losses.CategoricalCrossentropy())

model.build(input_shape=(2, 1))
# Print results before training.
print(model.predict(X))
model.fit(X, Y)

# Print results after training.
print(model.predict(X))
```

If you run this, you get (output trimmed a bit)

```
[[0.33333334 0.33333334 0.33333334]
 [0.33333334 0.33333334 0.33333334]]

[[0.3335555  0.3335555  0.33288905]
 [0.3335555  0.3335555  0.33288905]]
```

so, you can see, the weights don't really change much during training (they do change a little).  But TensorFlow bias weights are initialized to 0, and since your input points themselves are 0, the activations are all 0 anyway.  If you change `X` to be created with `np.ones()` instead, you get this:

```
[[0.29163927 0.21785341 0.49050725]
 [0.29163927 0.21785341 0.49050725]]

[[0.29221147 0.21828084 0.48950768]
 [0.29221147 0.21828084 0.48950768]]
```

Now, the initialized layer weights (not biases) matter, since the input data points are not zeros.  You can see this gives different results.  And, you can also see that training really does not do much of anything.

So what happens if we initialize mlpack's weights to zeros?  You can do this by replacing the `FFN<> model` change with a specific initialization type: `FFN<NegativeLogLikelihood, ConstInitialization> model` (and that `ConstInitialization` will default to zeros).  If you do that, mlpack's prediction results before training are

```
   0.3333   0.3333
   0.3333   0.3333
   0.3333   0.3333
```

and after training they are:

```
   0.4921   0.4921
   0.4921   0.4921
   0.0158   0.0158
```

So in this sense ensmallen's default parameters for Adam seem to find a better solution than TensorFlow's.  I'm sure you could tune TensorFlow to do a better job of ruling out the third class too (and I'm also sure `ens::Adam` could be tuned further, and maybe other optimizers would do better).

Overall I don't see any particular bug here.  *Maybe* you could argue that mlpack's default initializers should initialize the biases to zero, and I don't have a hugely strong opinion on that either way, but even for this problem that doesn't really help---it just causes the network to initialize in a way where the problem is already solved.

Please let me know if I overlooked something, or can clarify any of my findings (or if there is a problem in any of my findings! :)).

**Comment by AllIAskOfYou** [comment]
Thank you @rcurtin. I am very grateful for your help and the time you took to check this things.

I understand all your points, however I believe you forgot to set some parameters (I may be wrong) to be able to truly compare the two results.
In the case of training with tensorflow you didn't set the epoch number when calling Fit(), so it was kept at the default of 1 epoch, therefore the change in results was small. If you set the epoch number to the default value that is used in adam optimizer from ensmallen (which is 100000), the learning will converge to basicly (0.5, 0.5, 0).

Another thing is, that I tried to use uniform random init with tensorflow for biases and the learning still converged. It took a little bit more epochs (let's say with zero init it was 100 to get good result and with random init it was 1000) but it still converged to the correct probabilites. When trying the same thing with mlpack the learning converged correctly only with zero init. When using random init it converged to nearly (0.25, 0.75, 0) which seems a little bit to specific for it to be just a coincidence.

And if that is just how it is, does this mean that the ensmallen optimizer is worse at getting out of local minimums?

I thank you in advance for your answears :)

The parameters I used are as follows:
- learning rate = 0.001,
- beta1 = 0.9
- beta2 = 0.999
- epsilon = 1e-7
- tolerance = 1e-9 (only in ensmallen)

- bias init = (in tensorflow alias for init are:) 'zero' and 'random_uniform'

- batch size = 2,
- epoch number = from 1 to 100000 depending on what i was testing

**Comment by AllIAskOfYou** [comment]
Okey, so I tried a few more things and figured out, that last part was truely a coincidence. If I change the tolerance to a smaller value, it's going to converge at about (0, 1, 0) or (1, 0, 0), so it kind off chooses one over the other. This still doesn't seem quite right to me. It truly seems to me as if i'ts having a hard time getting out of some local minimums. I'll try some more things tomorow.

**Comment by rcurtin** [comment]
My belief without working the math is that without regularization on the weights of each layer, the losses are equivalent for models that give `(p, 1 - p, 0)` for any `0 <= p <= 1`.  It will probably be a while until I can return to this.

**Comment by AllIAskOfYou** [comment]
Well, I don't think that is quite true. The formula for negative log loss for one example is:
$$L = -\sum_{i=1}^{n} t_i \cdot ln(p_i)$$
So in our case where we have two examples in a batch and $t_i$ is either 0 or 1 we get:
$$-ln(p_0^0) - ln(p_1^1)$$ and $$p_0^0+p_1^1=1$$ (since both examples have the same features - $p_i^0=p_i^1$ and we expect the third class probability to be 0)

We can now simply rewrite the loss as:
$$-ln(p) - ln(1-p)$$
The local extreme of the function is:
$$\frac{dL}{dp} = 0, \frac{2p-1}{p(1-p)} = 0, p=0.5$$
Left and right of p=0.5 the function rises to infinity, therfore at p=0.5 L has a minimum value.
So when the p converges to 0, or 1 (as is the case when using random init and small tolerance), it converges to the worst possible loss.
Maybe I'm missing something here but I don't know what.
Thanks again for your time.
